compiler/rustc_lint/src/dangling.rs

```diff
@@ -0,0 +1,227 @@
+use rustc_hir::{Block, Expr, ExprKind, HirId, LangItem};
+use rustc_middle::ty::{Ty, TyCtxt};
+use rustc_session::{declare_lint, impl_lint_pass};
+use rustc_span::symbol::sym;
+
+use crate::lints::InstantlyDangling;
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `dangling_pointers_from_temporaries` lint detects getting a pointer to data
+    /// of a temporary that will immediately get dropped.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// # unsafe fn use_data(ptr: *const u8) { }
+    /// fn gather_and_use(bytes: impl Iterator<Item = u8>) {
+    ///     let x: *const u8 = bytes.collect::<Vec<u8>>().as_ptr();
+    ///     unsafe { use_data(x) }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Getting a pointer from a temporary value will not prolong its lifetime,
+    /// which means that the value can be dropped and the allocation freed
+    /// while the pointer still exists, making the pointer dangling.
+    /// This is not an error (as far as the type system is concerned)
+    /// but probably is not what the user intended either.
+    ///
+    /// If you need stronger guarantees, consider using references instead,
+    /// as they are statically verified by the borrow-checker to never dangle.
+    pub DANGLING_POINTERS_FROM_TEMPORARIES,
+    Warn,
+    "detects getting a pointer from a temporary"
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum LifetimeExtension {
+    /// Lifetime extension has not kicked in yet, but it will soon.
+    /// Example: walking LHS of a function/method call.
+    EnableLater { after_exit: HirId, until_exit: HirId },
+    /// Lifetime extension is currently active.
+    /// Example: walking a function/method call's arguments.
+    Enable { until_exit: HirId },
+    /// Temporary disable lifetime extension.
+    /// Example: statements of a block that is a function/method call's argument.
+    Disable { until_exit: HirId },
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct DanglingPointers {
+    /// Trying to deal with argument lifetime extension.
+    ///
+    /// This produces a dangling pointer:
+    /// ```ignore (example)
+    /// let ptr = CString::new("hello").unwrap().as_ptr();
+    /// foo(ptr)
+    /// ```
+    ///
+    /// But this does not:
+    /// ```ignore (example)
+    /// foo(CString::new("hello").unwrap().as_ptr())
+    /// ```
+    ///
+    /// But this does:
+    /// ```ignore (example)
+    /// foo({ let ptr = CString::new("hello").unwrap().as_ptr(); ptr })
+    /// ```
+    ///
+    /// We have to deal with this situation somehow.
+    ///
+    /// If we were a visitor, we could just keep track of
+    /// when we enter and exit places where lifetime extension kicks in
+    /// during visiting/walking and update a boolean flag accordingly.
+    ///
+    /// But we are not a visitor. We are a LateLintPass.
```

Comment:
Nothing prevents us from implementing a visitor and calling it from the `LateLintPass`. Presumably we would kick a [`rustc_hir::intravisit::Visitor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html) from the `LateLintPass::check_fn` impl.

Reply:
Good idea! I had been thinking of doing that, but completely forgot about `check_fn` and though we would be doomed to call this on every `check_expr`.

Reply:
Unfortunately this does not play along well with `#[allow]` & `#[deny]` on statements inside fn. They are just ignored.

I tried making the visitor buffer the info for lints so that they can be emitted later in `check_expr`, but it seems like `check_expr`s get called *before* the `check_fn` or `check_body`.

Reply:
Done